### PR TITLE
fix(remote): use git for push to remote system

### PR DESCRIPTION
**Pull request by syu-w**
Call git push directly due to libgit2 bug

Also the token will be used mutiple times, so it cannot expire after 1 min.

Fixes: #4478

### CODE REVIEW

File: snapcraft/remote/git.py [hunk 1]
```diff
@@ -181,23 +184,49 @@ def push_url(
...
-                t.name
-                for t in self._repo.references.iterator(pygit2.GIT_REFERENCES_TAGS)
-            ]
-        else:
-            tag_refs = []
         try:
-            self._repo.remotes.create_anonymous(remote_url).push([refspec] + tag_refs)
-        except pygit2.GitError as error:
+            git_proc = subprocess.Popen(
+                ["git", "push", remote_url, refspec, "--tags", "--progress"],
```
Comment (sergiusens) [inline snapcraft/remote/git.py:189]: --tags needs to be conditional on push_tags being True

File: snapcraft/remote/git.py [hunk 2]
```diff
@@ -181,23 +184,49 @@ def push_url(
...
-            tag_refs = [
-                t.name
-                for t in self._repo.references.iterator(pygit2.GIT_REFERENCES_TAGS)
-            ]
-        else:
-            tag_refs = []
         try:
-            self._repo.remotes.create_anonymous(remote_url).push([refspec] + tag_refs)
-        except pygit2.GitError as error:
+            git_proc = subprocess.Popen(
```
Comment (sergiusens) [inline snapcraft/remote/git.py:188]: please add the github issue for pygit2 on why we need this

File: snapcraft/remote/git.py [hunk 3]
```diff
@@ -181,23 +184,66 @@ def push_url(
...
-        except pygit2.GitError as error:
+            with subprocess.Popen(
+                cmd,
+                cwd=str(self.path),
+                bufsize=1,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                universal_newlines=True,
+            ) as git_proc:
+                # do not block on reading from the pipes (not working on Windows)
```
Comment (mr-cal) [inline snapcraft/remote/git.py:203]: Can you rephrase this?

I'm not sure if it means:
1. You added this code so that it does work on windows
2. You added this code and it does not work on windows

If you mean option 1, then something like `set non-blocking to prevent read blocking on windows` may be more clear

File: snapcraft/remote/git.py [hunk 4]
```diff
@@ -181,23 +184,66 @@ def push_url(
             "Pushing %r to remote %r with refspec %r.", ref, stripped_url, refspec
         )
 
-        # Create a list of tags to push
+        # temporarily call git directly due to libgit2 bug that unable to push
+        # large repos using https. See https://github.com/libgit2/libgit2/issues/6385
```
Comment (mr-cal) [inline snapcraft/remote/git.py:188]: Will you also reference the snapcraft ticket?
https://github.com/snapcore/snapcraft/issues/4478

**Review by mr-cal — APPROVED** [review]
Nice fix!  I can now build the telegram snap with `remote-build`.

The progress looks really good in brief mode.  In verbose mode, it makes a lot of noise (see below). @tigarmo - do you have an opinion here?  Do you think we could improve this if we migrate remote-build to `craft-application` and use `craft-cli` instead of `logging`?

For example:
```
> snapcraft remote-build --verbosity=trace
...
If interrupted, resume with: 'snapcraft remote-build --recover --build-id snapcraft-telegram-desktop-d0218ebddeb6974367e7a8bf12a4323f'
Starting build
creating git repo: name=snapcraft-telegram-desktop-d0218ebddeb6974367e7a8bf12a4323f, owner=/~mr-cal, target=/~mr-cal
Sending build data to Launchpad: https://mr-cal:<token>@git.launchpad.net/~mr-cal/+git/snapcraft-telegram-desktop-d0218ebddeb6974367e7a8bf12a4323f/Sending build data to Launchpad: https://mr-cal:<token>@git.launchpad.net/~mr-cal/+git/snapcraft-telegram-desktop-d0218ebddeb6974367e7a8bf12a432…
Enumerating objects: 214093, done.
Counting objects:   0% (1/214093)
Counting objects:   1% (2141/214093)
Counting objects:   2% (4282/214093)
Counting objects:   3% (6423/214093)
Counting objects:   4% (8564/214093)
Counting objects:   5% (10705/214093)
Counting objects:   6% (12846/214093)
Counting objects:   7% (14987/214093)
Counting objects:   8% (17128/214093)
Counting objects:   9% (19269/214093)
Counting objects:  10% (21410/214093)
Counting objects:  11% (23551/214093)
Counting objects:  12% (25692/214093)
Counting objects:  13% (27833/214093)
Counting objects:  14% (29974/214093)
Counting objects:  15% (32114/214093)
Counting objects:  16% (34255/214093)
Counting objects:  17% (36396/214093)
Counting objects:  18% (38537/214093)
Counting objects:  19% (40678/214093)
Counting objects:  20% (42819/214093)
Counting objects:  21% (44960/214093)
Counting objects:  22% (47101/214093)
Counting objects:  23% (49242/214093)
Counting objects:  24% (51383/214093)
Counting objects:  25% (53524/214093)
Counting objects:  26% (55665/214093)
Counting objects:  27% (57806/214093)
...
```

File: snapcraft/remote/git.py (hunk 3 above)
Comment (syu-w) [reply snapcraft/remote/git.py:203]: I mean the `os.set_blocking()` is not supported in Windows until Python 3.12. So it will have exception or still have blocked I/O in Windows. Not sure if anyone actually use it.

File: snapcraft/remote/git.py (hunk 4 above)
Comment (syu-w) [reply snapcraft/remote/git.py:188]: I can add that too

File: snapcraft/remote/git.py [hunk 5]
```diff
@@ -181,23 +184,69 @@ def push_url(
...
+                if git_proc.stdout:
+                    os.set_blocking(git_proc.stdout.fileno(), False)
+                if git_proc.stderr:
+                    os.set_blocking(git_proc.stderr.fileno(), False)
+
+                while git_proc.poll() is None:
+                    if git_proc.stdout:
+                        git_stdout: str = git_proc.stdout.readline()
+                        if git_stdout:
+                            logger.info(git_stdout.strip())
```
Comment (lengau) [inline snapcraft/remote/git.py:216]: Do we want to eliminate indentation if git's output provides it? What about `rstrip` instead?

File: snapcraft/remote/git.py [hunk 6]
```diff
@@ -181,23 +184,69 @@ def push_url(
...
+                    if git_proc.stderr:
+                        git_stderr: str = git_proc.stderr.readline()
+                        if git_stderr:
+                            logger.error(git_stderr.strip())
```
Comment (lengau) [inline snapcraft/remote/git.py:220]: If we receive many lines to stdout or stderr at once, we probably want to output them all together.
```suggestion
                        while git_stderr: str = git_proc.stderr.readline():
                            logger.error(git_stderr.strip())
```

File: snapcraft/remote/git.py (hunk 5 above)
Comment (syu-w) [reply snapcraft/remote/git.py:216]: I don't know if git push had obvious indentations, but I can do that.

**Review by lengau — COMMENTED** [review]
Got a couple of UX questions

File: snapcraft/remote/git.py (hunk 6 above)
Comment (syu-w) [reply snapcraft/remote/git.py:220]: That's a good idea

File: snapcraft/remote/git.py [hunk 7]
```diff
@@ -181,23 +184,70 @@ def push_url(
...
+
+                while git_proc.poll() is None:
+                    if git_proc.stdout:
+                        while git_stdout := git_proc.stdout.readline():
+                            logger.info(git_stdout.rstrip())
+                    if git_proc.stderr:
+                        while git_stderr := git_proc.stderr.readline():
+                            logger.error(git_stderr.rstrip())
+                    # avoid too much looping, but not too slow to display progress
+                    time.sleep(0.01)
```
Comment (lengau) [inline snapcraft/remote/git.py:223]: re: @mr-cal's question about `craft-cli`, I think part of the trouble here is that `git push` by default outputs a lot of stuff to `stderr`. Maybe we want to either use `--porcelain` or not treat `stdout` and `stderr` differently (since git doesn't use `stderr` for error output)?

**Review by lengau — APPROVED** [review]
Looks good! I left a note about Callahan's stdout/stderr/ux stuff, but I don't have an opinion about what's better or worse

**Comment by mr-cal** [comment]
As a team, we discussed creating and landing this on a `hotfix/8.0` branch but I decided to land this on `main`.  We can decide which commits to cherry-pick from `main` in January.
